Description: GtkMenuButton: use :toggled instead of :clicked
Origin: upstream, https://git.gnome.org/browse/gtk+/commit/?id=10c3a2483635d1c1
Bug: https://bugs.launchpad.net/bugs/1562002
Last-Update: 2017-03-23

--- a/gtk/gtkmenubutton.c
+++ b/gtk/gtkmenubutton.c
@@ -133,7 +133,6 @@
   GtkArrowType arrow_type;
   gboolean use_popover;
   guint press_handled : 1;
-  guint in_click : 1;
 };
 
 enum
@@ -427,21 +426,15 @@
 }
 
 static void
-gtk_menu_button_clicked (GtkButton *button)
+gtk_menu_button_toggled (GtkToggleButton *button)
 {
   GtkMenuButton *menu_button = GTK_MENU_BUTTON (button);
   GtkMenuButtonPrivate *priv = menu_button->priv;
-  gboolean active;
-
-  if (priv->in_click)
-    return;
-
-  priv->in_click = TRUE;
+  gboolean active = gtk_toggle_button_get_active (button);
 
   if (priv->menu)
     {
-      active = !gtk_widget_get_visible (priv->menu);
-      if (active)
+      if (active && !gtk_widget_get_visible (priv->menu))
         {
           GdkEvent *event;
 
@@ -460,20 +453,14 @@
     }
   else if (priv->popover)
     {
-      active = !gtk_widget_get_visible (priv->popover);
       if (active)
         gtk_widget_show (priv->popover);
       else
         gtk_widget_hide (priv->popover);
     }
-  else
-    active = FALSE;
-
-  GTK_BUTTON_CLASS (gtk_menu_button_parent_class)->clicked (button);
 
-  gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (button), active);
-  gtk_toggle_button_toggled (GTK_TOGGLE_BUTTON (button));
-  priv->in_click = FALSE;
+  if (GTK_TOGGLE_BUTTON_CLASS (gtk_menu_button_parent_class)->toggled)
+    GTK_TOGGLE_BUTTON_CLASS (gtk_menu_button_parent_class)->toggled (button);
 }
 
 static void
@@ -506,7 +493,7 @@
   GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
   GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
   GtkContainerClass *container_class = GTK_CONTAINER_CLASS (klass);
-  GtkButtonClass *button_class = GTK_BUTTON_CLASS (klass);
+  GtkToggleButtonClass *toggle_button_class = GTK_TOGGLE_BUTTON_CLASS (klass);
 
   gobject_class->set_property = gtk_menu_button_set_property;
   gobject_class->get_property = gtk_menu_button_get_property;
@@ -517,7 +504,7 @@
   container_class->add = gtk_menu_button_add;
   container_class->remove = gtk_menu_button_remove;
 
-  button_class->clicked = gtk_menu_button_clicked;
+  toggle_button_class->toggled = gtk_menu_button_toggled;
 
   /**
    * GtkMenuButton:popup:
